### ISSUE TITLE
[documentation] WebHeaderCollection.Add(new string('a', 65536)) does not throw as documented

### ISSUE THREAD

**Issue by hughbe** [dotnet/runtime#1486]
From https://msdn.microsoft.com/en-us/library/y280s7kx(v=vs.110).aspx

> ArgumentOutOfRangeException | The length of value is greater than 65535.

From dotnet/corefx#20169:
```cs
public static IEnumerable<object[]> Add_Value_TestData()
{
    string longString = new string('a', 65536);
    yield return new object[] { longString, longString }
}

[Theory]
[MemberData(nameof(Add_Value_TestData))]
public void Add_ValidValue_Success(string value, string expectedValue)
{
    var headers = new WebHeaderCollection
    {
        { "name", value }
    };

    Assert.Equal(expectedValue, headers["name"]);
}
```

**Comment by karelz** [comment]
Same behavior on .NET Framework 4.8 --> documentation issue.

**Comment by karelz** [comment]
Based on code inspection and code experiment:
.NET Framework throws for `HttpListenerRespose.Headers`: https://referencesource.microsoft.com/#System/net/System/Net/WebHeaderCollection.cs,665
The value `WebHeaderCollectionType.HttpListenerResponse` is set only in `HttpListenerResponse`: https://referencesource.microsoft.com/#System/net/System/Net/WebHeaderCollection.cs,e89b4b499f50aee7,references

.NET Core does NOT throw in that situation. The code uses `WebHeaderCollectionType.WebResponse` instead:
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Net.WebHeaderCollection/src/System/Net/WebHeaderCollection.cs#L405

Which is NOT set anywhere: https://source.dot.net/#System.Net.WebHeaderCollection/System/Net/WebHeaderCollection.cs,a46b85cf83e97adc,references

**Comment by CarnaViire** [comment]
As Karel said, whether exception is thrown is determined based on a private field -- `m_type` for .NET Framework and `_type` for .NET Core -- and it throws only for a specific value, namely `m_Type==WebHeaderCollectionType.HttpListenerResponse` for .NET Framework and `_type == WebHeaderCollectionType.WebResponse` for .NET Core.

In .NET Framework `m_type` is set through an internal constructor https://referencesource.microsoft.com/#System/net/System/Net/WebHeaderCollection.cs,ea519ee7f5217ece,references and because of that it does represent the object that created the header collection as intended.

However, there's no such constructor in .NET Core (and it wasn't there ever since [the initial commit of the lib](https://github.com/dotnet/runtime/blob/3e859acf0f177b701b47a30f5ba0fe12f44969c1/src/libraries/System.Net.WebHeaderCollection/src/System/Net/WebHeaderCollection.cs#L542)). Though, `_type` value is still being set, but in very obscure implicit way. It is done inside private properties `AllowHttpRequestHeader` and `AllowHttpResponseHeader` -- since `_type` is `WebHeaderCollectionType.Unknown` by default, whichever property was called first will set it to `WebHeaderCollectionType.WebRequest` or `WebHeaderCollectionType.WebResponse` respectively.

These private properties are called from methods that have `HttpRequestHeader` or `HttpResponseHeader` enums in their parameters, e.g. `public void Remove(HttpResponseHeader header)`.

This implementation produces quite puzzling behavior, for example:
```c#
string longString = new string('a', 65536);
WebHeaderCollection headerCollection = new WebHeaderCollection();

// no exception, as _type == Unknown
headerCollection.Add("someHeader", longString); 

// implicitly changes _type to WebResponse
headerCollection.Add(HttpResponseHeader.SetCookie, "someValueToChangeType"); 

// fails with ArgumentOutOfRangeException, as _type == WebResponse now
headerCollection.Add("someHeader2", longString); 
```

I think it might be hard to explain this behavior in the docs. Maybe it will be better to either drop the `_type` altogether or just drop its checks for ArgumentOutOfRangeException?

**Comment by CarnaViire** [comment]
TL;DR: We have different behaviors (.NET Framework vs .NET Core) for both standalone usage of WebHeaderCollection and for usage of HttpListener/HttpListenerResponse. This might be a problem for people trying to migrate from .NET Framework to .NET Core.

----

Below is the difference of WebHeaderCollection behavior (.NET Framework vs .NET Core) for adding long headers in 3 situations:
1. Using square bracket operator
```c#
headerCollection["Long-Header"] = longString;
headerCollection[HttpResponseHeader.SetCookie] = "someValueToChangeType"; // this might implicitly change _type
headerCollection["Long-Header-2"] = longString;
```
2. Using Add function
```c#
headerCollection.Add("Long-Header", longString);
headerCollection.Add(HttpResponseHeader.SetCookie, "someValueToChangeType"); // this might implicitly change _type
headerCollection.Add("Long-Header-2", longString);
```
3. Using HttpListenerResponse (that contains WebHeaderCollection inside)
```c#
// WebHeaderCollection.Add(String,String) is called inside
response.AddHeader("Long-Header", longString); 

// WebHeaderCollection[HttpResponseHeader] is called inside
response.Redirect("someValueToChangeType"); // this might implicitly change _type

// WebHeaderCollection.Add(String,String) is called inside
response.AddHeader("Long-Header-2", longString); 
```

#### .NET Framework
Does not throw, unless it is inside HttpListenerResponse. In HttpListenerResponse, always throws.
```
[1] WebHeaderCollection_SquareBrackets
add Long-Header...success
possible _type change... success
add Long-Header-2... success
OK

[2] WebHeaderCollection_Add
add Long-Header... success
possible _type change... success
add Long-Header-2... success
OK

[3] HttpListenerResponse_AddHeader
add Long-Header... ERROR ArgumentOutOfRangeException
FAIL
```

#### .NET Core 1.0
Does not throw.
WebHeaderCollection.Add and HttpListener not implemented in .NET Core 1.0.

```
[1] WebHeaderCollection_SquareBrackets
add Long-Header...success
possible _type change... success
add Long-Header-2... success
OK
```

#### .NET Core 2.0 and later up to 5.0-preview
Throws after implicit _type change in all 3 situations.
```
[1] WebHeaderCollection_SquareBrackets
add Long-Header...success
possible _type change... success
add Long-Header-2... ERROR ArgumentOutOfRangeException
FAIL

[2] WebHeaderCollection_Add
add Long-Header... success
possible _type change... success
add Long-Header-2... ERROR ArgumentOutOfRangeException
FAIL

[3] HttpListenerResponse_AddHeader
add Long-Header... success
possible _type change... success
add Long-Header-2... ERROR ArgumentOutOfRangeException
FAIL
```

**Comment by CarnaViire** [comment]
As the current behavior is puzzling and potentially might be harmful, I believe we might want to fix it in some way. Below are possible fixes for that.

1. Add a public method to WebHeaderCollection that enables length check. Do the check based on that flag and not on _type.

We can't easily go back to the original behavior because WebHeaderCollection and HttpListenerResponse live in different libraries now, so we can't just add an internal method/constructor as before. Going through API review might not be what we want now. Still I am listing it as a potential fix option.

2. Never check for length.

RFC does not propose any limitations on header length. Our other implementation of headers in System.Net.Http does not have any length restrictions either https://docs.microsoft.com/en-us/dotnet/api/system.net.http.headers.httpheaders.add?view=netcore-3.1. I wasn't able to find any reasons for this check to be introduced and why this check might be useful now. Also, as in .NET Framework this validation is turned on only in 1 out of 13 WebHeaderCollection instantiations - it looks like the validation is mostly turned off in .NET Framework. So I believe the easiest fix would be to remove the validation.

3. Always check length and throw ArgumentOutOfRangeException.

I believe this is not a good option. This might be harmful for users that don't expect it.

----

JFYI below is the list of places where WebHeaderCollection is used now in .NET Core
- System.Net.FileWebRequest 
- System.Net.FileWebResponse
- System.Net.FtpWebRequest 
- System.Net.FtpWebResponse
- System.Net.HttpWebRequest
- System.Net.HttpWebResponse
- System.Net.HttpListenerRequest
- System.Net.HttpListenerResponse
- System.Net.WebClient
- System.Net.WebSockets.ClientWebSocketOptions
- Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs
